pkg/apis/experimental/types.go

```diff
@@ -639,3 +639,36 @@ type ClusterAutoscalerList struct {
 
 	Items []ClusterAutoscaler `json:"items"`
 }
+
+// Selector is a label query over a set of API objects. MatchLabels and
+// MatchExpressions are ANDed. An empty Selector matches all objects. A
+// null Selector matches no objects.
+type Selector struct {
+	// MatchLabels is a list of labels.
+	MatchLabels map[string]string `json:"matchLabels,omitempty"`
+	// MatchExpressions is a list of requirement selectors.
+	MatchExpressions []SelectorRequirement `json:"matchExpressions,omitempty"`
+}
+
+// SelectorRequirement is a selector that contains values, a key
+// and an operator that relates the key and values.
+type SelectorRequirement struct {
+	// Key is the label key that the selector applies to.
+	Key string `json:"key"`
+	// Operator represents a key's relationship
+	// to a set of values.
+	Operator SelectorOperator `json:"operator"`
+	// StringValues is a set of string values. If the operator is In or NotIn,
+	// the values set must be non-empty.
+	StringValues []string `json:"stringValues,omitempty"`
+}
+
+// SelectorOperator is the set of operators that can be used in
+// a selector requirement.
+type SelectorOperator string
+
+const (
+	SelectorOpIn     SelectorOperator = "In"
+	SelectorOpNotIn  SelectorOperator = "NotIn"
+	SelectorOpExists SelectorOperator = "Exists"
```

Comment:
Label selectors already have all the same race conditions as the escape hatches.

I'm not strongly opposed to Does Not Exist but it is not implemented now and I don't think there's strong justification for it based on use cases. 

Numeric comparison: You've argued that controllers shouldn't even have selectors, much less super-complicated selectors. 

I would like to restrict this PR to just what label selection we support in controllers and services. What we do for node selection and in the CLI and UI are a different story. I could totally imagine supporting more general filtering capabilities in tools. 

The more we add, the less likely we can do this for 1.1, in which case we'll either never move controllers to the more general selector, or we'd at least put it off for a very long time, since it would cause user pain.

Additionally, selection of pods in services and controllers has serious operational implications. This was literally discussed for months internally. Of the things you mentioned, regexps, at least, are definitely out.